### PR TITLE
5. change(db): Improve unspent transparent output storage and tests

### LINKED ISSUE #3952: Modify utxo_by_outpoint to utxo_by_out_loc in the finalized state
### Motivation

We want to answer transparent address RPCs quickly. So we need to make existing indexes in the finalized state smaller and more efficient.

We also need to return some RPCs in chain order.

### Designs

UTXOs are stored in `utxo_by_out_loc` by `OutputLocation`, rather than `OutPoint`. (This makes keys 28 bytes smaller.)

`OutPoint`s can be looked up using `tx_loc_by_hash`, and reconstructed using `hash_by_tx_loc`.

The `Utxo` type can be constructed from the `OutputLocation` and `Output` data:
- `height: OutputLocation.height`, and
- `is_coinbase: OutputLocation.transaction_index == 0` (coinbase transactions are always the first transaction in a block).

#### Types

- `OutputIndex`: 24 bits, big-endian, unsigned (max ~223,000 transfers in the 2 MB block limit)
   See `Height` serialization for an example implementation. 
- `OutputLocation`: `TransactionLocation \|\| OutputIndex`
- `AddressLocation`: the first `OutputLocation` used by a `transparent::Address`. Always has the same value for each address, even if the first output is spent.

https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0005-state-updates.md#rocksdb-data-structures

### Tasks

Implementation:
- [x] Modify the column family in the finalized state
    - [x] Change `OutputLocation` to use `OutputIndex` rather than `OutPoint`
    - [x] Store `Output` values rather than `Utxo` values (and delete `Utxo` serialization)
    - [x] Re-construct the `Utxo` from the `Output` value and `OutputLocation` key
    - [x] Store the `AddressLocation` value after the `Output` value
- [x] Increment the database version

Testing:
- [x] Update raw data snapshot tests
- [x] Update struct data snapshot tests

CI:
- [x] If you added new Rust module files, make sure the paths in the regenerate state job are up to date

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

To maintain the `utxo_by_transparent_addr_loc` when UTXOs are deleted, we need to store an address location with each UTXO in `utxo_by_out_loc`.

### Designs

The `AddressLocation` part of `utxo_by_out_loc` in:
https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/rfcs/0005-state-updates.md#rocksdb-data-structures

## Solution

Note: some of these changes are redundant, they get reverted in PR #3999.

- Add UnspentOutputAddressLocation
- Add the AddressLocation to the UTXO database value
- Increment the database version 

Related changes:
- Make AddressLocations clearer in the API

Test changes:
- Update the snapshot test code for UnspentOutputAddressLocation
- Update the raw data snapshots
- Update the high-level data snapshots
- Add test-only serialization, and make existing serialization test-only

I ran a full sync test at:
- https://github.com/ZcashFoundation/zebra/actions/runs/2056286854

It got to 98.730% after 6 hours, which is about the same performance.

Closes #3952.

## Review

This PR is on the critical path for the lightwalletd work, so I've tagged it as a high priority.

This is part of a PR series: #3978, #3992, and #3993.

This PR is based on PR #3992.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

## Follow Up Work

The rest of the lightwalletd database changes.

### CODE REVIEW

File: zebra-state/src/service/finalized_state/disk_format/tests/prop.rs [hunk 1]
```diff
@@ -165,6 +167,20 @@ fn roundtrip_transparent_output() {
     proptest!(|(val in any::<transparent::Output>())| assert_value_properties(val));
 }
 
+#[test]
+fn roundtrip_unspent_output_address_location() {
+    zebra_test::init();
+
+    proptest!(
+        |(mut val in any::<UnspentOutputAddressLocation>())| {
+            if let Some(address_location) =val.address_location() {
```
Comment (jvff) [inline zebra-state/src/service/finalized_state/disk_format/tests/prop.rs]: Minor formatting:

```suggestion
            if let Some(address_location) = val.address_location() {
```

File: zebra-state/src/service/finalized_state/disk_format/transparent.rs [hunk 2]
```diff
@@ -380,15 +448,49 @@ impl FromDisk for AddressBalanceLocation {
...
+    fn as_bytes(&self) -> Self::Bytes {
+        let output_bytes = self.output().zcash_serialize_to_vec().unwrap();
+
+        // If there is no address location, don't write any bytes for it
+        let address_location_bytes = self
+            .address_location()
+            .map(|address_location| address_location.as_bytes().to_vec())
+            .unwrap_or_default();
+
+        [output_bytes, address_location_bytes].concat()
```
Comment (jvff) [inline zebra-state/src/service/finalized_state/disk_format/transparent.rs]: Idea: this should eliminate an extra `Vec` allocation, but I'm not sure if:

a) it negatively impacts readability
b) the compiler optimizes this anyway

```suggestion
        let mut bytes = self.output().zcash_serialize_to_vec().unwrap();

        // If there is no address location, don't write any bytes for it
        if let Some(address_location) = self.address_location() {
            bytes.extend(address_location.as_bytes())
        }

        bytes
```

**Review by jvff — APPROVED** [review]
Looks great! I added one formatting nit-pick and one optional idea, but it can be merged without them :+1:

File: zebra-state/src/service/finalized_state/disk_format/transparent.rs (hunk 2 above)
Comment (teor2345) [reply zebra-state/src/service/finalized_state/disk_format/transparent.rs]: Thanks for the suggestion, I think it is clearer.

**Review by jvff — DISMISSED** [review]

**Review by gustavovalverde — APPROVED** [review]

**Comment by teor2345** [comment]
This PR is mostly reverted by the next PR in the series, so I'm going to close it, and merge the full set of commits in the next PR.

**Pull request closed without merging**